### PR TITLE
MODFEE-80 create waive check endpoint, add tests

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Approach
- create check-pay endpoint;
- implement validation;
- create tests;

Resolves: [MODFEE-80](https://issues.folio.org/browse/MODFEE-80)

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -209,6 +209,15 @@
             "accounts.pay.check.post"
           ]
         },
+        {
+          "methods":[
+            "POST"
+          ],
+          "pathPattern":"/accounts/{id}/check-waive",
+          "permissionsRequired":[
+            "accounts.waive.check.post"
```
Comment (alexanderkurash) [inline descriptors/ModuleDescriptor-template.json:218]: accounts.check-waive.post ?

File: ramls/accounts.raml [hunk 2]
```diff
@@ -66,6 +66,38 @@ resourceTypes:
...
+                example: "Internal server error"
+    /check-waive:
+      post:
+        description: Checks if waive is valid
+        body:
+          application/json:
+            type: accountsCheckRequest
+        responses:
+          200:
+            description: "Returns accounts check response if validation is successful"
```
Comment (alexanderkurash) [inline ramls/accounts.raml:92]: The style of this description is different from others. And, again, we're not checking the account itself.

File: ramls/accounts.raml [hunk 3]
```diff
@@ -66,6 +66,38 @@ resourceTypes:
...
+            body:
+              application/json:
+                type: accountsCheckResponse
```
Comment (alexanderkurash) [inline ramls/accounts.raml:95]: I'm not sure this is the best naming - we're not checking the account, we're checking the ability to perform an action against a particular account. Probably checkActionResponse would be better, but we also need to keep it consistent.

File: ramls/accountsCheckRequest.json [hunk 4]
```diff
@@ -4,8 +4,8 @@
   "type": "object",
   "properties": {
     "amount": {
-      "type": "number",
-      "description": "Amount of fine which patron pays"
+      "type": "string",
+      "description": "Amount of fine which patron filled in the form"
```
Comment (alexanderkurash) [inline ramls/accountsCheckRequest.json:8]: We shouldn't reference the UI specifics.

File: src/main/java/org/folio/rest/exception/FailedValidationException.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package org.folio.rest.exception;
+
+public class FailedValidationException extends RuntimeException{
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/exception/FailedValidationException.java:3]: Isn't this name too broad? Or it'll be used for any kind of validation errors?

File: src/test/java/org/folio/rest/utils/ResourceClients.java [hunk 6]
```diff
@@ -12,6 +12,10 @@ public static ResourceClient accountsPayCheckClient(String accountId) {
     return new ResourceClient("/accounts/" + accountId + "/check-pay");
   }
 
+  public static ResourceClient accountsWaiveCheckClient(String accountId) {
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/utils/ResourceClients.java:15]: Does it makes sense to change the URL's word ordering in this method's name?

File: src/main/java/org/folio/rest/impl/AccountsAPI.java [hunk 7]
```diff
@@ -324,54 +323,69 @@ public void putAccountsByAccountId(String accountId, String lang,
...
+    Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler,
+    Context vertxContext) {
+
+    String tenantId = TenantTool.calculateTenantId(okapiHeaders.get(OKAPI_HEADER_TENANT));
+    PostgresClient pgClient = PostgresClient.getInstance(vertxContext.owner(), tenantId);
+
+    FeeFineActionValidationService validationService = new FeeFineActionValidationService(
+      new AccountRepository(pgClient));
+    String rawAmount = request.getAmount();
+    validationService.validate(accountId, rawAmount)
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/AccountsAPI.java:352]: Blank line?

File: src/main/java/org/folio/rest/impl/AccountsAPI.java [hunk 8]
```diff
@@ -324,54 +323,69 @@ public void putAccountsByAccountId(String accountId, String lang,
...
-      AccountValidationService validationService = new AccountValidationService(
-        new AccountRepository(pgClient));
-      Double amount = request.getAmount();
+    validateAction(accountId, request, okapiHeaders, asyncResultHandler, vertxContext);
+  }
 
-      validationService.validate(accountId, amount)
-        .onSuccess(result -> {
-          if (result.isValid()) {
-            AccountsCheckResponse response = createBaseAccountCheckResponse(accountId, amount)
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/AccountsAPI.java:349]: CheckActionResponse?

File: src/main/java/org/folio/rest/impl/AccountsAPI.java [hunk 9]
```diff
@@ -324,54 +323,69 @@ public void putAccountsByAccountId(String accountId, String lang,
...
-              .withAllowed(false)
-              .withErrorMessage(((ActionValidationFailure) result).getErrorMessage());
-            asyncResultHandler.handle(Future.succeededFuture(
-              PostAccountsCheckPayByAccountIdResponse
-                .respond422WithApplicationJson(response)));
-          }
-        }).onFailure(e -> Future.succeededFuture(
-        PostAccountsCheckPayByAccountIdResponse.respond500WithTextPlain(e.getMessage())));
-    }
+  private void validateAction(String accountId, AccountsCheckRequest request,
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/AccountsAPI.java:342]: checkAction?

File: src/main/java/org/folio/rest/exception/FailedValidationException.java (hunk 5 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/exception/FailedValidationException.java:3]: yes, it's used for any kind of validation error e.g. "Invalid amount entered", "Requested amount exceeds remaining amount", "Amount must be positive"

File: descriptors/ModuleDescriptor-template.json (hunk 1 above)
Comment (roman-barannyk) [reply descriptors/ModuleDescriptor-template.json]: updated

File: ramls/accounts.raml (hunk 2 above)
Comment (roman-barannyk) [reply ramls/accounts.raml]: fixed

File: ramls/accounts.raml (hunk 3 above)
Comment (roman-barannyk) [reply ramls/accounts.raml]: renamed

File: ramls/accountsCheckRequest.json (hunk 4 above)
Comment (roman-barannyk) [reply ramls/accountsCheckRequest.json]: rephrased

File: src/test/java/org/folio/rest/utils/ResourceClients.java (hunk 6 above)
Comment (roman-barannyk) [reply src/test/java/org/folio/rest/utils/ResourceClients.java]: reordered

File: src/main/java/org/folio/rest/impl/AccountsAPI.java (hunk 7 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/impl/AccountsAPI.java:352]: Sorry, I didn't get that. Why the blank line should be here?

File: src/main/java/org/folio/rest/impl/AccountsAPI.java (hunk 8 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/impl/AccountsAPI.java:349]: updated

File: src/main/java/org/folio/rest/impl/AccountsAPI.java (hunk 9 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/impl/AccountsAPI.java]: renamed

File: ramls/accounts.raml [hunk 10]
```diff
@@ -59,18 +59,50 @@ resourceTypes:
...
               application/json:
-                type: accountsCheckResponse
+                type: checkActionResponse
+          404:
+            description: "Account was not found"
+            body:
+              text/plain:
+                example: "Account was not found"
+          422:
+            description: "Returns accounts pay check response with validation error"
```
Comment (alexanderkurash) [inline ramls/accounts.raml:75]: Payment is not allowed

File: ramls/accounts.raml [hunk 11]
```diff
@@ -59,18 +59,50 @@ resourceTypes:
...
+              application/json:
+                type: checkActionResponse
+          404:
+            description: "Account was not found"
+            body:
+              text/plain:
+                example: "Account was not found"
           422:
-            description: "Returns accounts check response with validation error"
+            description: "Returns accounts waive check response with validation error"
```
Comment (alexanderkurash) [inline ramls/accounts.raml:102]: Waiver is not allowed

File: ramls/checkActionRequest.json [hunk 12]
```diff
@@ -4,8 +4,8 @@
   "type": "object",
   "properties": {
     "amount": {
-      "type": "number",
-      "description": "Amount of fine which patron pays"
+      "type": "string",
+      "description": "Amount of fine which patron fulfills"
```
Comment (alexanderkurash) [inline ramls/checkActionRequest.json:8]: Action amount

File: ramls/accounts.raml [hunk 13]
```diff
@@ -59,18 +59,50 @@ resourceTypes:
         description: Checks if payment is valid
         body:
           application/json:
-            type: accountsCheckRequest
+            type: checkActionRequest
         responses:
           200:
-            description: "Returns accounts check response if validation is successful"
+            description: "Returns accounts pay check response if validation is successful"
```
Comment (alexanderkurash) [inline ramls/accounts.raml:65]: Payment is allowed

File: src/test/java/org/folio/rest/impl/AccountsActionChecksAPITests.java [hunk 14]
```diff
@@ -16,64 +19,168 @@
   @Test
   public void payCheckAmountShouldBeAllowed() {
     Account accountToPost = postAccount();
-    ResourceClient accountsPayCheckClient = accountsPayCheckClient(accountToPost.getId());
-    AccountsCheckRequest accountCheckRequest = new AccountsCheckRequest();
-    accountCheckRequest.withAmount(3.0);
+    ResourceClient accountsPayCheckClient = accountsCheckPayClient(accountToPost.getId());
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/AccountsActionChecksAPITests.java:22]: Rename accountsPayCheckClient

File: src/test/java/org/folio/rest/impl/AccountsActionChecksAPITests.java [hunk 15]
```diff
@@ -16,64 +19,168 @@
...
+  }
 
-    accountsPayCheckClient.attemptCreate(accountCheckRequest)
-      .then()
-      .statusCode(HttpStatus.SC_OK)
-      .body("allowed", is(true))
-      .body("amount", is(new Float(accountCheckRequest.getAmount())))
-      .body("remainingAmount", is(new Float(accountToPost.getRemaining())));
+  @Test
+  public void waiveCheckAmountShouldBeAllowed() {
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/AccountsActionChecksAPITests.java:27]: checkWaive... ?

File: ramls/accounts.raml (hunk 10 above)
Comment (roman-barannyk) [reply ramls/accounts.raml]: done

File: ramls/accounts.raml (hunk 11 above)
Comment (roman-barannyk) [reply ramls/accounts.raml]: done

File: ramls/checkActionRequest.json (hunk 12 above)
Comment (roman-barannyk) [reply ramls/checkActionRequest.json:8]: done

File: ramls/accounts.raml (hunk 13 above)
Comment (roman-barannyk) [reply ramls/accounts.raml]: done

File: src/main/java/org/folio/rest/impl/AccountsAPI.java [hunk 16]
```diff
@@ -324,54 +323,69 @@ public void putAccountsByAccountId(String accountId, String lang,
...
-        PostAccountsCheckPayByAccountIdResponse.respond500WithTextPlain(e.getMessage())));
-    }
+  private void checkAction(String accountId, CheckActionRequest request,
+    Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler,
+    Context vertxContext) {
+
+    String tenantId = TenantTool.calculateTenantId(okapiHeaders.get(OKAPI_HEADER_TENANT));
+    PostgresClient pgClient = PostgresClient.getInstance(vertxContext.owner(), tenantId);
+
+    FeeFineActionValidationService validationService = new FeeFineActionValidationService(
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/impl/AccountsAPI.java]: renamed

File: src/main/java/org/folio/rest/impl/AccountsAPI.java [hunk 17]
```diff
@@ -324,54 +323,69 @@ public void putAccountsByAccountId(String accountId, String lang,
...
-                .respond422WithApplicationJson(response)));
-          }
-        }).onFailure(e -> Future.succeededFuture(
-        PostAccountsCheckPayByAccountIdResponse.respond500WithTextPlain(e.getMessage())));
-    }
+  private void checkAction(String accountId, CheckActionRequest request,
+    Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler,
+    Context vertxContext) {
+
+    String tenantId = TenantTool.calculateTenantId(okapiHeaders.get(OKAPI_HEADER_TENANT));
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/impl/AccountsAPI.java]: updated

File: src/test/java/org/folio/rest/impl/AccountsActionChecksAPITests.java (hunk 14 above)
Comment (roman-barannyk) [reply src/test/java/org/folio/rest/impl/AccountsActionChecksAPITests.java]: renamed

File: src/test/java/org/folio/rest/impl/AccountsActionChecksAPITests.java (hunk 15 above)
Comment (roman-barannyk) [reply src/test/java/org/folio/rest/impl/AccountsActionChecksAPITests.java]: renamed

**Pull request merged**